### PR TITLE
Verify that remote target bucket is on MinIO server for bucket replication 

### PR DESCRIPTION
**Pull request by taran-p**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Adds an initial health check when adding a remote target for bucket replication. This ensures the target is a MinIO server.

## Motivation and Context
Non-MinIO servers are allowed to be added as remote targets, which is not supported

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by taran-p** [comment]
@harshavardhana This is a quick replacement for minio/mc#4747 , I just added a health check on the server side rather than on `mc`'s side. Should this be hitting an authenticated API instead or does this work?

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Is there some code missing ? I don't see any relevant code

File: cmd/bucket-targets.go [hunk 1]
```diff
@@ -351,6 +351,25 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
 			return BucketRemoteTargetNotVersioned{Bucket: tgt.TargetBucket}
 		}
 	}
+
+	// Check if target is a MinIO server and alive
+	hcCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
```
Comment (harshavardhana) [inline cmd/bucket-targets.go]: This is too long, keep it like 3 secs, it is more than enough. Anywhere where latency is bigger than 3 seconds then it's a unusable cluster anyway.

We can even indicate to the user that their remote target doesn't respond to the expected 3 sec response time.

This way they know what is expected from a replication target.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by taran-p** [comment]
Changed the timing and error. Is there anything else to clarify?

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

File: cmd/bucket-targets.go [hunk 2]
```diff
@@ -351,6 +351,25 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
...
+	result := <-sys.hcClient.Alive(hcCtx, madmin.AliveOpts{}, madmin.ServerProperties{
+		Endpoint: tgt.Endpoint,
+		Scheme:   scheme,
+	})
+	cancel()
+	if result.Error != nil {
+		return RemoteTargetConnectionErr{Bucket: tgt.TargetBucket, Err: result.Error, AccessKey: tgt.Credentials.AccessKey}
+	}
+	if !result.Online {
+		return BucketRemoteTargetNotAlive{Bucket: tgt.TargetBucket}
```
Comment (vadmeste) [inline cmd/bucket-targets.go]: It seems to me you don't have to define a new error type, you can still use RemoteTargetConnectionErr for simplicity.

**Review by vadmeste — COMMENTED** [review]
One comment

File: cmd/bucket-targets.go [hunk 3]
```diff
@@ -351,6 +351,25 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
...
 		}
 	}
+
+	// Check if target is a MinIO server and alive
+	hcCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	scheme := "http"
+	if tgt.Secure {
+		scheme = "https"
+	}
+	result := <-sys.hcClient.Alive(hcCtx, madmin.AliveOpts{}, madmin.ServerProperties{
```
Comment (vadmeste) [inline cmd/bucket-targets.go:362]: It looks like madmin.ServerProperties is not needed as well; hcClient has already the URL to ping.

Comment (taran-p) [reply cmd/bucket-targets.go:362]: hcClient would not have been given the URL of the target replication bucket. The function did not work properly when testing without the ServerProperties

**Review by poornas — APPROVED** [review]

**Pull request merged**